src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs

```diff
@@ -57,103 +61,159 @@ public DateHeaderValueManager()
+            //Are we in a new second?
+            if (now - lastReadDateTime >= TimeSpan.FromSeconds(1) || now.Second != lastReadDateTime.Second)
+            {
+                //Yep, Update DateValues
+                SetDateValues(now);
+                WriteLongThreadSafe(ref _lastReadDateTimeTicks, now.Ticks);
+            }
 
             if (_hadRequestsSinceLastTimerTick)
             {
                 // We served requests since the last tick, reset the flag and return as we're still active
                 _hadRequestsSinceLastTimerTick = false;
-                _lastRequestSeen = now;
+                WriteLongThreadSafe(ref _lastRequestSeenTicks, now.Ticks);
                 return;
             }
 
             // No requests since the last timer tick, we need to check if we're beyond the idle threshold
-            var timeSinceLastRequestSeen = now - _lastRequestSeen;
-            if (timeSinceLastRequestSeen >= _timeWithoutRequestsUntilIdle)
+            if ((now.Ticks - ReadLongThreadSafe(ref _lastRequestSeenTicks)) >= _timeWithoutRequestsUntilIdle.Ticks)
             {
                 // No requests since idle threshold so stop the timer if it's still running
-                if (_dateValueTimer != null)
+                StopTimer();
+            }
+        }
+
+        /// <summary>
+        /// Starts the timer if it's turned off, or sets the datevalues to the current time if disposed. 
+        /// </summary>
+        private void PrepareDateValues()
+        {
+            if (_isDisposed)
+            {
+                SetDateValues(_systemClock.UtcNow);
+            }
+            else
+            {
+                if (!_timerIsRunning)
                 {
-                    lock (_timerLocker)
-                    {
-                        if (_dateValueTimer != null)
-                        {
-                            DisposeTimer();
-                        }
-                    }
+                    StartTimer();
                 }
             }
         }
 
-        private void DisposeTimer()
+
+        /// <summary>
+        /// Sets date values from a provided ticks value
+        /// </summary>
+        /// <param name="value">A DateTimeOffset value</param>
+        private void SetDateValues(DateTimeOffset value)
+        {
+            // See http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.18 for required format of Date header
+            _dateValue = value.ToString(Constants.RFC1123DateFormat);
+            Encoding.ASCII.GetBytes(_dateValue, 0, _dateValue.Length, !_activeDateBytes ? _dateBytes0 : _dateBytes1, "\r\nDate: ".Length);
+            _activeDateBytes = !_activeDateBytes;
+        }
+
+        private void WriteLongThreadSafe(ref long location, long value)
         {
-            if (_dateValueTimer != null)
+            Interlocked.Exchange(ref location, value);
+        }
+
+        private long ReadLongThreadSafe(ref long location)
```

Comment:
Yes, Interlocked.Read is mostly useful for 32-bit systems. It eliminates the possibility of having a torn read, which can't happen when reading a 64 bit value on a 64 bit machine (since .NET aligns the word properly).